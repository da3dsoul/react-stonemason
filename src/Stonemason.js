import React, { useState, useLayoutEffect, useRef} from 'react';
import PropTypes from 'prop-types';
import ResizeObserver from 'resize-observer-polyfill';
import { computeRowLayout } from './layout';

const Stonemason = function Stonemason({margin, limitNodeSearch, targetRowHeight, children}) {
  const [containerWidth, setContainerWidth] = useState(0);
  const StonemasonEl = useRef(null);
  let photos = children;

  useLayoutEffect(() => {
    let animationFrameID = null;
    const observer = new ResizeObserver(entries => {
      // only do something if width changes
      const newWidth = entries[0].contentRect.width;
      if (containerWidth !== newWidth) {
        // put in an animation frame to stop "benign errors" from
        // ResizeObserver https://stackoverflow.com/questions/49384120/resizeobserver-loop-limit-exceeded
        animationFrameID = window.requestAnimationFrame(() => {
          setContainerWidth(Math.floor(newWidth));
        });
      }
    });
    observer.observe(StonemasonEl.current);
    return () => {
      observer.disconnect();
      window.cancelAnimationFrame(animationFrameID);
    };
  });

  // no containerWidth until after first render with refs, skip calculations and render nothing
  if (!containerWidth) return <div ref={StonemasonEl}>&nbsp;</div>;
  // subtract 1 pixel because the browser may round up a pixel
  const width = containerWidth - 1;
  const StonemasonStyle = { display: 'flex', flexWrap: 'wrap', flexDirection: 'row' };
  photos = computeRowLayout({ containerWidth: width, limitNodeSearch, targetRowHeight, margin, photos });

  return (
    <div className="react-stonemason--Stonemason">
      <div ref={StonemasonEl} style={StonemasonStyle}>
        {photos}
      </div>
    </div>
  );
};

const imagePropType = PropTypes.shape({
  key: PropTypes.string,
  props: PropTypes.shape({
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired
  })
});

Stonemason.propTypes = {
  children: PropTypes.arrayOf(imagePropType).isRequired,
  targetRowHeight: PropTypes.oneOfType([PropTypes.func, PropTypes.number]),
  limitNodeSearch: PropTypes.oneOfType([PropTypes.func, PropTypes.number]),
  margin: PropTypes.number
};

Stonemason.defaultProps = {
  margin: 2,
  targetRowHeight: 300,
};
export default Stonemason;
