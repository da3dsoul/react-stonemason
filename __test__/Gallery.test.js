import React from 'react';
import Stonemason from '../src/Stonemason';
import { photos } from './test-photo-data';
import { mount } from 'enzyme';

describe('Stonemason', () => {
  let wrapper;
  afterEach(() => {
    if (wrapper && wrapper.length > 0) {
      wrapper.unmount();
    }
  });

  it('it matches correct snapshot', () => {
    wrapper = mount((
        <Stonemason>
          {photos.map(a => <img key={a.key} title={a.title} alt={a.alt} src={a.src} srcSet={a.srcSet} width={a.width} height={a.height} sizes={a.sizes} />)}
        </Stonemason>));
    expect(wrapper).toMatchSnapshot();
  });
});
