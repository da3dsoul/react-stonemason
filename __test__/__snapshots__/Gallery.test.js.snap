// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stonemason it matches correct snapshot 1`] = `
<Stonemason
  margin={2}
  targetRowHeight={300}
>
  <div
    className="react-stonemason--Stonemason"
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "flexWrap": "wrap",
        }
      }
    >
      <img
        alt="DSC02576"
        height={1065}
        key="40673436833"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/40673436833_5fc1ab52fe_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/40673436833_5fc1ab52fe.jpg 500w",
            "https://live.staticflickr.com/65535/40673436833_5fc1ab52fe_c.jpg 800w",
            "https://live.staticflickr.com/65535/40673436833_5fc1ab52fe_b.jpg 1024w",
            "https://live.staticflickr.com/65535/40673436833_6f245db6aa_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "250.44444444444446px",
            "margin": "2px",
            "width": "375.7px",
          }
        }
        title="DSC02576"
        width={1600}
      />
      <img
        alt="DSC02565"
        height={1065}
        key="47587238132"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/47587238132_4683fdabb2_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/47587238132_4683fdabb2.jpg 500w",
            "https://live.staticflickr.com/65535/47587238132_4683fdabb2_c.jpg 800w",
            "https://live.staticflickr.com/65535/47587238132_4683fdabb2_b.jpg 1024w",
            "https://live.staticflickr.com/65535/47587238132_54f98ac5e9_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "250.44444444444446px",
            "margin": "2px",
            "width": "375.7px",
          }
        }
        title="DSC02565"
        width={1600}
      />
      <img
        alt="DSC02548"
        height={1065}
        key="32697497947"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/32697497947_7d6dfaf5ec_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/32697497947_7d6dfaf5ec.jpg 500w",
            "https://live.staticflickr.com/65535/32697497947_7d6dfaf5ec_c.jpg 800w",
            "https://live.staticflickr.com/65535/32697497947_7d6dfaf5ec_b.jpg 1024w",
            "https://live.staticflickr.com/65535/32697497947_81c6d9139e_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "250.44444444444446px",
            "margin": "2px",
            "width": "375.7px",
          }
        }
        title="DSC02548"
        width={1600}
      />
      <img
        alt="DSC02540"
        height={1065}
        key="32697497647"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/32697497647_31cd88c0b5_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/32697497647_31cd88c0b5.jpg 500w",
            "https://live.staticflickr.com/65535/32697497647_31cd88c0b5_c.jpg 800w",
            "https://live.staticflickr.com/65535/32697497647_31cd88c0b5_b.jpg 1024w",
            "https://live.staticflickr.com/65535/32697497647_7a36437b05_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "250.44444444444446px",
            "margin": "2px",
            "width": "375.7px",
          }
        }
        title="DSC02540"
        width={1600}
      />
      <img
        alt="DSC02535"
        height={1065}
        key="47587238242"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/47587238242_6776b8a582_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/47587238242_6776b8a582.jpg 500w",
            "https://live.staticflickr.com/65535/47587238242_6776b8a582_c.jpg 800w",
            "https://live.staticflickr.com/65535/47587238242_6776b8a582_b.jpg 1024w",
            "https://live.staticflickr.com/65535/47587238242_feab7e0aa0_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "250.44444444444446px",
            "margin": "2px",
            "width": "375.7px",
          }
        }
        title="DSC02535"
        width={1600}
      />
      <img
        alt="DSC02520"
        height={1065}
        key="40673437163"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/40673437163_703ffe416d_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/40673437163_703ffe416d.jpg 500w",
            "https://live.staticflickr.com/65535/40673437163_703ffe416d_c.jpg 800w",
            "https://live.staticflickr.com/65535/40673437163_703ffe416d_b.jpg 1024w",
            "https://live.staticflickr.com/65535/40673437163_7f4a4d66f2_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "250.44444444444446px",
            "margin": "2px",
            "width": "375.7px",
          }
        }
        title="DSC02520"
        width={1600}
      />
      <img
        alt="DSC02494"
        height={1065}
        key="47587238362"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/47587238362_5d355937cb_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/47587238362_5d355937cb.jpg 500w",
            "https://live.staticflickr.com/65535/47587238362_5d355937cb_c.jpg 800w",
            "https://live.staticflickr.com/65535/47587238362_5d355937cb_b.jpg 1024w",
            "https://live.staticflickr.com/65535/47587238362_65fe216f4b_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "332.64705882352945px",
            "margin": "2px",
            "width": "499px",
          }
        }
        title="DSC02494"
        width={1600}
      />
      <img
        alt="DSC02492"
        height={840}
        key="32697498187"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/32697498187_7a74849c76_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/32697498187_7a74849c76.jpg 500w",
            "https://live.staticflickr.com/65535/32697498187_7a74849c76_c.jpg 800w",
            "https://live.staticflickr.com/65535/32697498187_7a74849c76_b.jpg 1024w",
            "https://live.staticflickr.com/65535/32697498187_9d07b79e86_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "332.64705882352945px",
            "margin": "2px",
            "width": "632px",
          }
        }
        title="DSC02492"
        width={1600}
      />
      <img
        alt="DSC02489"
        height={1065}
        key="40673436913"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/40673436913_c48ac9e6d5_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/40673436913_c48ac9e6d5.jpg 500w",
            "https://live.staticflickr.com/65535/40673436913_c48ac9e6d5_c.jpg 800w",
            "https://live.staticflickr.com/65535/40673436913_c48ac9e6d5_b.jpg 1024w",
            "https://live.staticflickr.com/65535/40673436913_dbb2320a41_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "241.32762312633832px",
            "margin": "2px",
            "width": "362px",
          }
        }
        title="DSC02489"
        width={1600}
      />
      <img
        alt="DSC02479"
        height={957}
        key="40673437213"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/40673437213_3480e88c8b_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/40673437213_3480e88c8b.jpg 500w",
            "https://live.staticflickr.com/65535/40673437213_3480e88c8b_c.jpg 800w",
            "https://live.staticflickr.com/65535/40673437213_3480e88c8b_b.jpg 1024w",
            "https://live.staticflickr.com/65535/40673437213_5bfff3ef95_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "241.32762312633832px",
            "margin": "2px",
            "width": "403px",
          }
        }
        title="DSC02479"
        width={1600}
      />
      <img
        alt="DSC02461"
        height={1065}
        key="47587238482"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/47587238482_5ac51de790_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/47587238482_5ac51de790.jpg 500w",
            "https://live.staticflickr.com/65535/47587238482_5ac51de790_c.jpg 800w",
            "https://live.staticflickr.com/65535/47587238482_5ac51de790_b.jpg 1024w",
            "https://live.staticflickr.com/65535/47587238482_b8b1478145_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "241.32762312633832px",
            "margin": "2px",
            "width": "362px",
          }
        }
        title="DSC02461"
        width={1600}
      />
      <img
        alt="DSC02460"
        height={1065}
        key="32697498227"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/32697498227_a3bc50aa6c_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/32697498227_a3bc50aa6c.jpg 500w",
            "https://live.staticflickr.com/65535/32697498227_a3bc50aa6c_c.jpg 800w",
            "https://live.staticflickr.com/65535/32697498227_a3bc50aa6c_b.jpg 1024w",
            "https://live.staticflickr.com/65535/32697498227_e5a478e65c_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "252.1252796420582px",
            "margin": "2px",
            "width": "378.2px",
          }
        }
        title="DSC02460"
        width={1600}
      />
      <img
        alt="DSC02456"
        height={1086}
        key="40673437263"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/40673437263_b03be98bae_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/40673437263_b03be98bae.jpg 500w",
            "https://live.staticflickr.com/65535/40673437263_b03be98bae_c.jpg 800w",
            "https://live.staticflickr.com/65535/40673437263_b03be98bae_b.jpg 1024w",
            "https://live.staticflickr.com/65535/40673437263_bebc91464c_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "252.1252796420582px",
            "margin": "2px",
            "width": "370.6px",
          }
        }
        title="DSC02456"
        width={1600}
      />
      <img
        alt="DSC02453"
        height={1065}
        key="32697497717"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/32697497717_1fb25b32be_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/32697497717_1fb25b32be.jpg 500w",
            "https://live.staticflickr.com/65535/32697497717_1fb25b32be_c.jpg 800w",
            "https://live.staticflickr.com/65535/32697497717_1fb25b32be_b.jpg 1024w",
            "https://live.staticflickr.com/65535/32697497717_4bf3235964_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "252.1252796420582px",
            "margin": "2px",
            "width": "378.2px",
          }
        }
        title="DSC02453"
        width={1600}
      />
      <img
        alt="DSC02441"
        height={1065}
        key="47587238552"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/47587238552_cfce31cc10_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/47587238552_cfce31cc10.jpg 500w",
            "https://live.staticflickr.com/65535/47587238552_cfce31cc10_c.jpg 800w",
            "https://live.staticflickr.com/65535/47587238552_cfce31cc10_b.jpg 1024w",
            "https://live.staticflickr.com/65535/47587238552_44cd3655ea_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "250.44444444444446px",
            "margin": "2px",
            "width": "375.7px",
          }
        }
        title="DSC02441"
        width={1600}
      />
      <img
        alt="DSC02439"
        height={1065}
        key="47640016631"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/47640016631_cfffe28dd0_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/47640016631_cfffe28dd0.jpg 500w",
            "https://live.staticflickr.com/65535/47640016631_cfffe28dd0_c.jpg 800w",
            "https://live.staticflickr.com/65535/47640016631_cfffe28dd0_b.jpg 1024w",
            "https://live.staticflickr.com/65535/47640016631_ec4ec9f250_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "250.44444444444446px",
            "margin": "2px",
            "width": "375.7px",
          }
        }
        title="DSC02439"
        width={1600}
      />
      <img
        alt="DSC02436"
        height={1065}
        key="47640017061"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/47640017061_d6f554a175_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/47640017061_d6f554a175.jpg 500w",
            "https://live.staticflickr.com/65535/47640017061_d6f554a175_c.jpg 800w",
            "https://live.staticflickr.com/65535/47640017061_d6f554a175_b.jpg 1024w",
            "https://live.staticflickr.com/65535/47640017061_cf1fd70288_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "250.44444444444446px",
            "margin": "2px",
            "width": "375.7px",
          }
        }
        title="DSC02436"
        width={1600}
      />
      <img
        alt="DSC02435"
        height={708}
        key="32697497787"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/32697497787_ebd9165050_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/32697497787_ebd9165050.jpg 500w",
            "https://live.staticflickr.com/65535/32697497787_ebd9165050_c.jpg 800w",
            "https://live.staticflickr.com/65535/32697497787_ebd9165050_b.jpg 1024w",
            "https://live.staticflickr.com/65535/32697497787_e3c6cb4bc5_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "254.4018058690745px",
            "margin": "2px",
            "width": "574.9px",
          }
        }
        title="DSC02435"
        width={1600}
      />
      <img
        alt="DSC02428"
        height={1600}
        key="40673437303"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/40673437303_ec8b5a7246_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/40673437303_ec8b5a7246.jpg 333w",
            "https://live.staticflickr.com/65535/40673437303_ec8b5a7246_c.jpg 533w",
            "https://live.staticflickr.com/65535/40673437303_ec8b5a7246_b.jpg 682w",
            "https://live.staticflickr.com/65535/40673437303_c88d6174a2_h.jpg 1065w",
          ]
        }
        style={
          Object {
            "height": "254.4018058690745px",
            "margin": "2px",
            "width": "170.4px",
          }
        }
        title="DSC02428"
        width={1065}
      />
      <img
        alt="DSC02427"
        height={1065}
        key="46724351385"
        sizes="(min-width: 480px) 50vw, (min-width: 1024px) 33.3vw, 100vw"
        src="https://live.staticflickr.com/65535/46724351385_a6ff50ddda_b.jpg"
        srcSet={
          Array [
            "https://live.staticflickr.com/65535/46724351385_a6ff50ddda.jpg 500w",
            "https://live.staticflickr.com/65535/46724351385_a6ff50ddda_c.jpg 800w",
            "https://live.staticflickr.com/65535/46724351385_a6ff50ddda_b.jpg 1024w",
            "https://live.staticflickr.com/65535/46724351385_3a29cde51c_h.jpg 1600w",
          ]
        }
        style={
          Object {
            "height": "254.4018058690745px",
            "margin": "2px",
            "width": "381.6px",
          }
        }
        title="DSC02427"
        width={1600}
      />
    </div>
  </div>
</Stonemason>
`;
